### PR TITLE
Fix ENGINE method redirection.

### PR DESCRIPTION
**Pull request by snhenson**
If we are passed an ENGINE to use in int_ctx_new e.g. via EVP_PKEY_CTX_new()
use it instead of the default.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

This PR will address some bugs/deficiencies when the EVP_PKEY framework is used to redirect operations to an ENGINE: currently it doesn't work properly. This will address the problems of #4495.

### CODE REVIEW

**Comment by snhenson** [comment]
This doesn't address everything yet but there is one significant bug fix already: we clobbered the passed engine when calling EVP_PKEY_CTX_new().

@mouse07410 this should partially fix the EVP_DigestSignInit() problems you were seeing: if you now pass an explicit EVP_PKEY_CTX into EVP_DigestSignInit() created using EVP_PKEY_CTX_new() with the appropriate EVP_PKEY and ENGINE it _should_ now redirect the operations to your ENGINE instead of using the defaults.

That's only a part fix though. An application shouldn't need to do that. If that works I'll work on the correct fix.

**Comment by snhenson** [comment]
Update: added a function EVP_PKEY_set1_engine() if you call this after setting up the key (set type, key components etc) it should redirect operation to the ENGINE at the EVP_PKEY level. That should happen transparently and no need to specify an ENGINE parameter to EVP_DigesSignInit() etc.

Documentation and tests to follow.

**Comment by snhenson** [comment]
Note: I have some ugly test code that generates a dummy EVP_PKEY_METHOD and checks redirection. With these updates it seems to work as expected now. I'll work on tidying that up and including it.

**Comment by mouse07410** [comment]
> Update: added a function `EVP_PKEY_set1_engine()` if you call this after setting up the key...

Before I've tried anything: I'd *dearly* love if you could also add functions

- EVP_PKEY_get_engine()
- EVP_PKEY_CTX_get_engine()
- EVP_PKEY_CTX_set1_engine()

**Comment by snhenson** [comment]
Why would you need those? I'm not saying they shouldn't be added I'm just seeing if the required functionality is already there. For example EVP_PKEY_CTX_set1_engine() would appear to be similar to EVP_PKEY_CTX_new() (with the fix in this PR) but without the complications of what to do if an ENGINE/method is swapped part way through an operation.

**Comment by mouse07410** [comment]
> Why would you need those?

The setters - to be able to change things post-factum. The getters - to be able to learn what I was given (what I'm dealing with), and possibly pass the `ENGINE *` over to other components.

**Comment by mouse07410** [comment]
OK, first report. For  PIV tokens it worked fine for both `RSA-PKCS` and `RSA-PSS`.

For the HSM, it seemed to do do `RSA-PKCS` fine, but failed on `RSA-PSS`:
```
. . . . .
OSSL-Tst5/main.c:673 sign: failed to sign (rv=0)
140736337826752:error:8207A070:PKCS#11 module:pkcs11_private_encrypt:Mechanism invalid:p11_rsa.c:140:
Signing failed! (rv=0 olen=384)
```
Symptoms are indicative of trying to do everything in software, offloading to the token only the last raw RSA operation.

**Comment by snhenson** [comment]
OK I'll add my test when it's in a usable state: that does redirect properly.

**Comment by mouse07410** [comment]
More details. Trying to breakpoint at the entrance of `EVP_DigestSignInit()` I observe that `pkey->engine == NULL` (actually, all the `->engine` references I checked down the `EVP_DigestSignFinal()` were NULL). Here is my test/demo code (error checks removed for brevity):
```
int sign(const unsigned char *in, const size_t inlen, unsigned char **out,
         size_t *olen, ENGINE *e)
{
    int            rv      = -1;
    int            saltlen = -2; /* max permitted salt len value */
    EVP_PKEY      *privkey = NULL;
    EVP_PKEY_CTX  *ctx     = NULL;
    EVP_MD_CTX    *md_ctx  = NULL;
    const EVP_MD  *md      = HASH;
    const int      padding = RSA_PKCS1_PSS_PADDING;
    // SignPrivKey is a pkcs11: URI pointing at the right key (with RSA signing capabilities) in the HSM

    PW_CB_DATA     cb_data;
    cb_data.password = NULL;
    cb_data.prompt_info = "PIN:";
    
    /* Ensure engine pointer is valid */
    
    privkey = ENGINE_load_private_key(e, SignPrivKey, NULL, &cb_data);
    md_ctx = EVP_MD_CTX_create();
 
    rv = EVP_DigestSignInit(md_ctx, &ctx, md, NULL, privkey);
    rv = EVP_DigestSignUpdate(md_ctx, in, inlen)
    // set_signature_md() has been done internally in DigestSignInit()
    rv = EVP_PKEY_CTX_set_rsa_padding(ctx, padding);
    rv = EVP_PKEY_CTX_set_rsa_mgf1_md(ctx, md);
    rv =  EVP_PKEY_CTX_set_rsa_pss_saltlen(ctx, saltlen);
    *olen = 0;
    rv = EVP_DigestSignFinal(md_ctx, NULL, olen);
    *out = OPENSSL_malloc(*olen);
    rv = EVP_DigestSignFinal(md_ctx, *out, olen);    

    return rv;
}
```

And here's an excerpt from PKCS#11 trace, showing that OpenSSL attempts to perform the (forbidden by HSM) raw RSA operation:
```
199: C_SignInit
2017-10-09 16:13:27.679
[in] hSession = 0x0
pMechanism->type=CKM_RSA_X_509                
[in] hKey = 0x30301
Returned:  112 CKR_MECHANISM_INVALID
```

**Comment by snhenson** [comment]
You need to explicitly indicate that an ENGINEs EVP_PKEY_METHOD will handle a particular key with EVP_PKEY_set1_engine(): OpenSSL can't work that out for itself. You'd typically call that in the ENGINE itself in its load_private_key routine, if appropriate.

I just added a test for redirection which uses that technique.

**Comment by mouse07410** [comment]
> You need to explicitly indicate that an ENGINE's `EVP_PKEY_METHOD` will handle a particular key with `EVP_PKEY_set1_engine()`: OpenSSL can't work that out for itself. You'd typically call that in the ENGINE itself in its load_private_key routine, if appropriate.

Sorry, I'm not sure I fully understand.

I had code that did not know any of that. It did not work for tokens that do internal padding. I added one line `EVP_PKEY_set1_engine()` that forced/set the engine (again?) for the EVP_PKEY object. That fixed the problem, and the code worked OK with no more changes. I did not do anything with `EVP_PKEY_METHOD`, beyond what pkcs11 engine is already doing - but this engine is invariant, as in both tests (with and without `EVP_PKEY_set1_engine()` call before `EVP_DigestSignInit()`) it was the same.

You seem to suggest that the call `EVP_PKEY_set1_engine()` should be done within the engine in its method that loads private keys? You also said "if appropriate". How would engine (or I) know when it is or isn't appropriate? Depending on what? As I said, I don't think I fully understand...

**Comment by snhenson** [comment]
When an ENGINE returns a key in its load_privkey() routine it typically has to tell OpenSSL which ENGINE handles that key. At the RSA level it will do this explicitly via a call to RSA_new_method(engine). This point of EVP_PKEY_set1_engine() is to do the same but at the EVP_PKEY level instead of the RSA_METHOD level.

By "appropriate" I mean an ENGINE will call EVP_PKEY_set1_engine() if (a) the ENGINE contains an EVP_PKEY_METHOD for the key algorithm and (b) the key it is returning should use that method.

As for applicability of this PR to other branches. Well at least _some_ of it is clear cut bug fixes. I'd argue that not being able to redirect operations to an ENGINE at the EVP_PKEY level is another bug too but that's just my opinion. For 1.0.2 EVP_PKEY is not opaque and the EVP_PKEY_set1_engine() changes the EVP_PKEY structure so that's problematic: though it could be added to the end.

**Comment by mouse07410** [comment]
> By "appropriate" I mean an ENGINE will call EVP_PKEY_set1_engine() if (a) the ENGINE contains an EVP_PKEY_METHOD for the key algorithm and (b) the key it is returning should use that method.

Thank you, now I understand (at least I hope I do ;).

Would you object to adding this call to `crypto/engine/eng_pkey.c`, something like
```
diff --git a/crypto/engine/eng_pkey.c b/crypto/engine/eng_pkey.c
index 305a648feb..629d591a25 100644
--- a/crypto/engine/eng_pkey.c
+++ b/crypto/engine/eng_pkey.c
@@ -78,6 +78,7 @@ EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                   ENGINE_R_FAILED_LOADING_PRIVATE_KEY);
         return 0;
     }
+    EVP_PKEY_set1_engine(pkey, e);
     return pkey;
 }
```
The current PKCS#11 engine seems able to determine whether the operation should be completely offloaded to the engine, or done in software (except the case when data digest is involved ;). I've tested this change, and it appears to work fine. 

> I'd argue that not being able to redirect operations to an ENGINE at the EVP_PKEY level is another bug too but that's just my opinion.

I for one concur 100% with your opinion. I suspect there would be quite a few of other like-minded people, especially among those who use HSM (or other tokens that can do padding internally) and probably PIV tokens.

> For 1.0.2 EVP_PKEY is not opaque and the EVP_PKEY_set1_engine() changes the EVP_PKEY structure so that's problematic: though it could be added to the end.

For 1.0.2 should the engine pointer just forcefully override `ENGINE *engine` member of the structure? After all, inside the `load_privkey()` engine method that member can hardly carry anything (else) of value?

And a semi-generic question. I'd like the engine code to use `EVP_PKEY_set1_engine()`, but it is unclear when I could expect every OpenSSL installation to be patched to offer it. So, how can I check from within the engine, whether this method is or isn't available, preferably at compile time, but fine if during run-time?  Thanks!

**Comment by snhenson** [comment]
I can't really add that call because it might not be applicable to the ENGINE in question. If, for example, the ENGINE redirects operations at the RSA_METHOD level but _not_ EVP_PKEY_METHOD it **must not** call EVP_PKEY_set1_engine() and that call will return an error. So the ENGINE has to decide that for itself.

**Comment by mouse07410** [comment]
> I can't really add that call because it might not be applicable to the ENGINE in question.

Understood, thanks. A pity, though. ;-)

Is there a good straightforward way for an engine (or other code) to determine whether `EVP_PKEY_set1_engine()` is available in the given binary OpenSSL installation?

**Comment by snhenson** [comment]
Well since it adds a new function code you could test to see if EVP_F_EVP_PKEY_SET1_ENGINE is defined.

**Comment by mouse07410** [comment]
OK, now with the engine with `EVP_PKEY_set1_engine()` added to its `load_privkey()` my test-code seems to work fine (on PIV and HSM, with RSA-PKCS and RSA-PSS). So it looks like the majority of the issues have been nailed.

How would we proceed to having this fix merged to master and percolated to 1.1.0?

**Comment by snhenson** [comment]
Update: rebased against master and documentation added.

**Comment by snhenson** [comment]
OK this should be ready for review now. 

The 1.1.0 backport of EVP_PKEY_set1_engine() is debatable as to whether it is a new feature or not... I'd argue that it is a bug fix because without this you can't use EVP_PKEY_METHOD in an ENGINE properly (as bit like a missing accessor). 

Alternatively I can just keep this as master for now and have the less controversial bug fixes for backport to 1.1.0 which would be a separate PR.

**Comment by mouse07410** [comment]
> The 1.1.0 backport of `EVP_PKEY_set1_engine()` is debatable as to whether it is a new feature or not... I'd argue that it is a bug fix because without this you can't use `EVP_PKEY_METHOD` in an ENGINE properly (as bit like a missing accessor).

You have a better access to the OpenSSL dev team than I do, so you know better when the above can be discussed - and maybe even can predict how the discussion would go. But in my humble opinion, how can this be considered as anything *but* a missing accessor case? A missing accessor that makes it impossible to use combined modes like `EVP_DigestSignInit()` with engines/tokens that insist on applying padding like PSS internally (rather than "outsourcing" it to the OpenSSL software)?

If you think it would hold the progress of this PR for too long, then probably your alternative strategy of applying this to master, and submitting a second (similar) PR for 1.1.0 is a better option. I can't gauge the best course because I don't know what to expect from the rest of the team.

File: test/enginetest.c [hunk 1]
```diff
@@ -175,6 +176,177 @@ static int test_engines(void)
...
+    EVP_PKEY *pk;
+
+    rsa = RSA_new();
+
+    pk = EVP_PKEY_new();
+
+    if (rsa == NULL || pk == NULL)
+        return NULL;
+
+    EVP_PKEY_assign_RSA(pk, rsa);
```
Comment (mattcaswell) [inline test/enginetest.c]: Shouldn't we test the return value here?

File: test/enginetest.c [hunk 2]
```diff
@@ -175,6 +176,177 @@ static int test_engines(void)
...
+    len = EVP_PKEY_size(pkey);
+    if (!TEST_ptr(tmp = OPENSSL_malloc(len)))
+        goto err;
+
+    if (!TEST_ptr(ctx = EVP_PKEY_CTX_new(pkey, NULL)))
+        goto err;
+    TEST_info("EVP_PKEY_encrypt test: no redirection");
+    /* Encrypt some data: should succeed but not be redirected */
+    if (!TEST_true(EVP_PKEY_encrypt_init(ctx) > 0)
+            || !TEST_true(EVP_PKEY_encrypt(ctx, tmp, &len, pt, sizeof(pt)) > 0)
```
Comment (mattcaswell) [inline test/enginetest.c]: TEST_int_gt() instead of TEST_true() in these two lines and other places in this function?

File: test/enginetest.c [hunk 3]
```diff
@@ -175,6 +176,177 @@ static int test_engines(void)
...
+    /* Setup an empty test EVP_PKEY_METHOD and set callback to return it */
+    if (!TEST_ptr(test_rsa = EVP_PKEY_meth_new(EVP_PKEY_RSA, 0)))
+        goto err;
+    ENGINE_set_pkey_meths(e, test_pkey_meths);
+
+    /* Getting a context for test ENGINE should now succeed */
+    if (!TEST_ptr(ctx = EVP_PKEY_CTX_new(pkey, e)))
+        goto err;
+    /* Encrypt should fail because operation is not supported */
+    if (!TEST_true(EVP_PKEY_encrypt_init(ctx) <= 0))
```
Comment (mattcaswell) [inline test/enginetest.c]: TEST_int_le()? And other places in this function

File: crypto/include/internal/evp_int.h [hunk 4]
```diff
@@ -369,6 +369,7 @@ struct evp_pkey_st {
     CRYPTO_REF_COUNT references;
     const EVP_PKEY_ASN1_METHOD *ameth;
     ENGINE *engine;
+    ENGINE *pmeth_engine; /* If not NULL public key ENGINE to use */
```
Comment (mattcaswell) [inline crypto/include/internal/evp_int.h]: I am slightly confused by the existing `engine` member of this structure and how it differs from `pmeth_engine`

Comment (snhenson) [reply crypto/include/internal/evp_int.h]: The existing `engine` pointer is set if the ASN.1 operations are associated with an ENGINE and is then also used for public key operations. An example of this is the GOST ENGINE which performs both ASN.1 and public key operations in an ENGINE. There will be cases where an ENGINE performs public key operations only and ASN.1 operations are handled either by OpenSSL itself or a different ENGINE. That's what `pmeth_engine` is for.

File: test/enginetest.c (hunk 1 above)
Comment (snhenson) [reply test/enginetest.c]: Done.

File: test/enginetest.c (hunk 2 above)
Comment (snhenson) [reply test/enginetest.c]: Addressed this and all the other cases (unless I missed one) in latest commit.

File: test/enginetest.c [hunk 5]
```diff
@@ -175,6 +176,174 @@ static int test_engines(void)
...
+            || !TEST_true(called_encrypt))
+        goto err;
+
+    EVP_PKEY_CTX_free(ctx);
+    ctx = NULL;
+    called_encrypt = 0;
+
+    /* Create context with default engine: should not be redirected */
+    ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!TEST_int_gt(EVP_PKEY_encrypt_init(ctx), 0)
```
Comment (mattcaswell) [inline test/enginetest.c]: Should there be a `TEST_ptr(ctx)` here?

File: test/enginetest.c [hunk 6]
```diff
@@ -175,6 +176,174 @@ static int test_engines(void)
...
+
+    /* Set engine explicitly for test key */
+    if (!TEST_true(EVP_PKEY_set1_engine(pkey, e)))
+        goto err;
+
+    TEST_info("EVP_PKEY_encrypt test: redirection via EVP_PKEY_set1_engine()");
+
+    /* Create context with default engine: should be redirected now */
+    ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!TEST_int_gt(EVP_PKEY_encrypt_init(ctx), 0)
```
Comment (mattcaswell) [inline test/enginetest.c]: And here.

**Comment by snhenson** [comment]
OK just to be crystal clear on this. Does the approval cover the 1.1.0 backport too or should I put that in a separate PR?

File: util/libcrypto.num [hunk 7]
```diff
@@ -4400,3 +4400,4 @@ EVP_PKEY_meth_remove                    4343	1_1_1	EXIST::FUNCTION:
 OPENSSL_sk_reserve                      4344	1_1_1	EXIST::FUNCTION:
 CRYPTO_atomic_read                      4345	1_1_1	EXIST::FUNCTION:
 CRYPTO_atomic_write                     4346	1_1_1	EXIST::FUNCTION:
+EVP_PKEY_set1_engine                    4347	1_1_1	EXIST::FUNCTION:ENGINE
```
Comment (mattcaswell) [inline util/libcrypto.num]: Needs to be 1_1_0g

Comment (snhenson) [reply util/libcrypto.num]: Yes and the error codes need to be consistent too.

**Comment by snhenson** [comment]
Pushed now to master and 1.1.0 apart from the enginetest where the changes were so extensive it deserved a separate PR see #4520

**Comment by mouse07410** [comment]
Thanks!

**Comment by mouse07410** [comment]
@snhenson I'm having problems with OpenSSL-1.0.2, as it does not provide `EVP_PKEY_set1_engine()`.  Can it be added there? Since there seems to be an agreement that this absence is a bug?

Because we not only need the ability to access the structure - but to differentiate between `pkey->engine` and `pkey->key_engine`, which the current `OpenSSL_1_0_2-stable` lacks.

**Comment by levitte** [comment]
Sorry, this is not possible for 1.0.2.  One of the reasons is that the `EVP_PKEY` structure is public instead of opaque there, and adding another field is not backward compatible.

**Comment by mouse07410** [comment]
> Sorry, this is not possible for 1.0.2...

@levitte OK, I understand. But since it is a bug (that interferes with combined mechanisms like `EVP_DigestSignInit()`), perhaps we can resurrect https://github.com/openssl/openssl/pull/4495 specifically for 1.0.2?

**Pull request closed without merging**